### PR TITLE
Fixing Stat Overflows consideration on DQM modules

### PR DESCRIPTION
**Pull request by jfernan2**
#### PR description:

This PR tries to fix the problem mentioned in https://github.com/cms-sw/cmssw/pull/32343#issuecomment-736862320 

The proposed solution goes directly to DQM core setStatOverflows method instead of using teh ROOT method SetStatOverflow since DQM Monitor Elements inherit from there and this prevents modifying many DQM modules around.

I let coding experts to decide if this solution is preferred to directly modify 
DQM/SiStripMonitorCluster/src/SiStripMonitorCluster.cc
DQM/SiStripMonitorTrack/src/SiStripMonitorTrack.cc
DQM/SiStripMonitorDigi/src/SiStripMonitorDigi.cc
DQMOffline/EGamma/src/ElectronDqmAnalyzerBase.cc
as suggested on https://github.com/cms-sw/cmssw/pull/3234 instead.

I can switch to that strategy if preferred.

I have also noticed that there are many Alignment classes suffering from this too:
https://github.com/cms-sw/cmssw/search?p=1&q=StatOverflows

#### PR validation:

Checked compilation and some wf running

### CODE REVIEW

**Comment by jfernan2** [comment]
please test

**Comment by mmusich** [comment]
@jfernan2 

> I have also noticed that there are many Alignment classes suffering from this too:
https://github.com/cms-sw/cmssw/search?p=1&q=StatOverflows

none of those (Alignment classes and macros) is run in central workflows, so I think they cannot do any harm anywhere else.
About:

```
DQM/SiStripMonitorCluster/src/SiStripMonitorCluster.cc
DQM/SiStripMonitorTrack/src/SiStripMonitorTrack.cc
DQM/SiStripMonitorDigi/src/SiStripMonitorDigi.cc
```

I think there is an ongoing campaign to remove  every direct access to the underlying bare ROOT from the ME within the tracker DQM group (see e.g: [this talk]( https://indico.cern.ch/event/934813/contributions/4103235/attachments/2142915/3611297/Removal%20of%20TH1%28%29_TkDQM_13Nov.pdf) )
Wouldn't solution at https://github.com/cms-sw/cmssw/pull/32343 going into the opposite direction?
(I am genuinely asking, as I am not expert)
@sroychow @arossi83

**Comment by jfernan2** [comment]
Thanks @mmusich  I am open to implement the other solution, I do not which one is best

**Comment by mmusich** [comment]
>  I do not know which one is best.

IMHO the one you propose here is the better one, but I am definitely not an expert.

**Comment by jfernan2** [comment]
please test
Let's see if this solution works without changes

**Comment by silviodonato** [comment]
> Comparison is ready
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-13b23a/11253/summary.html
> 
> Comparison Summary:
> 
>     * No significant changes to the logs found
> 
>     * Reco comparison results: 4 differences found in the comparisons
> 
>     * DQMHistoTests: Total files compared: 35
> 
>     * DQMHistoTests: Total histograms compared: 2529593
> 
>     * DQMHistoTests: Total failures: 7
> 
>     * DQMHistoTests: Total nulls: 0
> 
>     * DQMHistoTests: Total successes: 2529564
> 
>     * DQMHistoTests: Total skipped: 22
> 
>     * DQMHistoTests: Total Missing objects: 0
> 
>     * DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
> 
>     * Checked 148 log files, 37 edm output root files, 35 DQM output files

@jfernan2 there are no actual changes here, as you expected

File: DQMServices/Core/src/MonitorElement.cc [hunk 1]
```diff
@@ -906,7 +906,10 @@ namespace dqm::impl {
 
   void MonitorElement::setStatOverflows(unsigned int value) {
     auto access = this->accessMut();
-    access.value.object_->StatOverflows(value);
+    if (value == kTRUE)
+      access.value.object_->StatOverflows(TH1::kConsider);
```
[the remark this answers is not in the document]
Comment (jfernan2) [reply DQMServices/Core/src/MonitorElement.cc:910]: I believe you are right, I was thinking it was an internal DQM method. Thanks!

**Comment by jfernan2** [comment]
Hi @silviodonato 
Let's see if the last commit produces changes, however I would like to know the opinion of the experts before proceeding with this PR, becase I see two ways of doing this as mentioned in previous comments

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
Now there are indeed many modifications because the overflows are actually considered:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_3_X_2020-12-14-1100+13b23a/40467/dqm-histo-comparison-summary.html
Specially in Pixel MEs, @mmusich what do you think? Thanks

**Comment by mmusich** [comment]
@jfernan2 

>  what do you think? Thanks

I think that's what it is. If that's acceptable or not I can't say, I would let @sroychow and @arossi83 to have a look.
Can this be put on hold in the meanwhile?

To further clarify my main comment above: the approach used in #32343 seems to be in direct contradiction with earlier recommendations from the central DQM group.

**Comment by jfernan2** [comment]
please hold

> To further clarify my main comment above: the approach used in #32343 seems to be in direct contradiction with earlier recommendations from the central DQM group.
No DQM modules were involved in that PR

**Comment by jfernan2** [comment]
hold

**Comment by missirol** [comment]
> the approach used in #32343 seems to be in direct contradiction with earlier recommendations from the central DQM group.

One clarification from my side (for what it's worth): I was not aware of those DQM recommendations (@jfernan2 , could you please point to them?). What was done in #32343 was the only way, at that time, to avoid using ROOT's `StatOverflows` (I think using `SetStatOverflows` is the better way, since it doesn't cause changes to the outputs of unrelated modules). Of course, if a call to `SetStatOverflows` becomes available via `MonitorElement` (as done in this PR), there will be a followup to #32343 to use that instead of doing it via `getTH1()`.

**Comment by jfernan2** [comment]
@missirol   in principle DQM methods are working on internal "ROOT" objects:
https://github.com/cms-sw/cmssw/blob/d2c2ace7ad9f675712281e72b2f865a514c32aef/DQMServices/Core/interface/MonitorElement.h#L370-L430
so, to avoid colliding modifications which are difficult to track, direct ROOT modifications are not desired

**Comment by jfernan2** [comment]
ping @sroychow @arossi83

**Comment by mmusich** [comment]
@cmsbuild please test

**Comment by mmusich** [comment]
@jfernan2, 
Alessandro (@arossi83) is trying to reproduce locally but doens't seem to see the same things as the bot. I am retriggering the tests in order to get a fresh comparison.

**Comment by arossi83** [comment]
@jfernan2 

We investigate a little bit the difference we see, in particular on PixelPhase1 plots.
The differences are due to the fact that now the SetStatsOverflows applies only to the specific ME and not on the Global behavior of the ROOT object as before. Due to this PixelPhase1 plots are by default with StatsOverflows disabled.

I introduce a quick&dirty patch to verify this by manually enabling StatsOverflows for PixelPhase1 plots. In attachment you can find a plots with overlayed this PR, the baseline and the PR+Patch. You can see that by re-enabling the StatsOverflows no changes arise for Phase1 plots.
Test_StatOverflow

In light of that the hold can be removed from this PR so that it can be merged. We will fix the TrackerDQM part with a separate PR.

**Comment by jfernan2** [comment]
Thank you very much @arossi83 !!

**Comment by jfernan2** [comment]
unhold

**Comment by jfernan2** [comment]
@amassiro  would you be so kind to perform the followup to #32343 you mentioned below? Thanks

> One clarification from my side (for what it's worth): I was not aware of those DQM recommendations (@jfernan2 , could you please point to them?). What was done in #32343 was the only way, at that time, to avoid using ROOT's `StatOverflows` (I think using `SetStatOverflows` is the better way, since it doesn't cause changes to the outputs of unrelated modules). Of course, if a call to `SetStatOverflows` becomes available via `MonitorElement` (as done in this PR), there will be a followup to #32343 to use that instead of doing it via `getTH1()`.

**Comment by missirol** [comment]
> @amassiro would you be so kind to perform the followup to #32343 you mentioned below? Thanks

(i guess the message was for me.. :))

Yes, will do, as soon as this is merged.

**Comment by jfernan2** [comment]
Oh yes, sorry, github misled me. Thank you @missirol !

**Comment by jfernan2** [comment]
+1

**Comment by qliphy** [comment]
+1

**Pull request merged**